### PR TITLE
Add TLS/SSL channel close timeouts

### PR DESCRIPTION
**Pull request by Tim-Brooks**
Closing a channel using TLS/SSL requires reading and writing a
CLOSE_NOTIFY message (for pre-1.3 TLS versions). Many implementations do
not actually send the CLOSE_NOTIFY message, which means we are depending
on the TCP close from the other side to ensure channels are closed. In
case there is an issue with this, we need a timeout. This commit adds a
timeout to the channel close process for TLS secured channels.

As part of this change, we need a timer service. We could use the
generic Elasticsearch timeout threadpool. However, it would be nice to
have a local to the nio event loop timer service dedicated to network needs. In 
the future this service could support read timeouts, connect timeouts,
request timeouts, etc. This commit adds a basic priority queue backed
service. Since our timeout volume (channel closes) is very low, this
should be fine. However, this can be updated to something more efficient
in the future if needed (timer wheel). Everything being local to the event loop
thread makes the logic simple as no locking or synchronization is necessary.

### CODE REVIEW

**Comment by tbrooks8** [comment]
run cradle build tests 2

**Comment by tbrooks8** [comment]
run gradle build tests 2

File: libs/nio/src/main/java/org/elasticsearch/nio/NioTimer.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    long nanosUntilNextTask(long relativeNanos) {
+        DelayedTask nextTask = tasks.peek();
+        if (nextTask == null) {
+            return Long.MAX_VALUE;
+        } else {
+            return Math.max(nextTask.deadline - relativeNanos, 0);
+        }
+    }
+
+    public interface Cancellable {
```
Comment (s1monw) [inline libs/nio/src/main/java/org/elasticsearch/nio/NioTimer.java]: can we just use Runnable instead?

File: libs/nio/src/main/java/org/elasticsearch/nio/NioTimer.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+/**
+ * A basic priority queue backed timer service. The service is thread local and should only be used by a
+ * single nio selector event loop thread.
+ */
+public class NioTimer {
+
+    private final PriorityQueue<DelayedTask> tasks = new PriorityQueue<>(Comparator.comparingLong(DelayedTask::getDeadline));
+
+    public Cancellable schedule(Runnable task, TimeValue timeValue) {
```
Comment (s1monw) [inline libs/nio/src/main/java/org/elasticsearch/nio/NioTimer.java]: I think this is trappy, can we only have the `scheduleAtRelativeTime` method instead?

File: libs/nio/src/main/java/org/elasticsearch/nio/NioTimer.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        return scheduleAtRelativeTime(task, System.nanoTime() + timeValue.nanos());
+    }
+
+    public Cancellable scheduleAtRelativeTime(Runnable task, long relativeNanos) {
```
Comment (s1monw) [inline libs/nio/src/main/java/org/elasticsearch/nio/NioTimer.java]: javadocs please

File: libs/nio/src/main/java/org/elasticsearch/nio/NioTimer.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+import org.elasticsearch.common.unit.TimeValue;
+
+import java.util.Comparator;
+import java.util.PriorityQueue;
+
+/**
+ * A basic priority queue backed timer service. The service is thread local and should only be used by a
+ * single nio selector event loop thread.
+ */
+public class NioTimer {
```
Comment (s1monw) [inline libs/nio/src/main/java/org/elasticsearch/nio/NioTimer.java]: I think the `Nio` in the name is obsolete?

**Review by s1monw — APPROVED** [review]
I left some comments. looks good though. I was pretty confused about the use of `ThreadLocal` in your description. Maybe you can change that?

**Comment by tbrooks8** [comment]
run gradle build tests 1

**Pull request merged**